lxd/instance/drivers/driver_common.go

```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
 		return err
 	}
 
+	// Snapshot root disk.
 	err = pool.CreateInstanceSnapshot(snap, inst, d.op)
 	if err != nil {
 		return fmt.Errorf("Create instance snapshot: %w", err)
 	}
 
+	// Snapshot attached disk volumes (if requested).
+	if diskVolumesMode == api.DiskVolumesModeAllExclusive {
+		// Freeze instance for crash-consistent snapshots.
+		if inst.IsRunning() && !inst.IsFrozen() {
+			// Freeze the processes.
+			err = inst.Freeze()
+			if err != nil {
+				return err
+			}
+
+			defer func() { _ = inst.Unfreeze() }()
+
+			// Attempt to sync the filesystem.
+			err = filesystem.SyncFS(inst.Path())
+			if err != nil {
+				d.logger.Warn("Failed to flush writes to instance volume", logger.Ctx{"err": err})
+			}
+		}
+
+		volumes, err := d.getAttachedVolumes(inst)
+		if err != nil {
+			return fmt.Errorf("Failed getting attached volumes: %w", err)
+		}
+
+		// Snapshot attached custom volumes.
+		storageCache := storagePools.NewStorageCache(pool) // Create storage cache for pool lookups.
+		attachedVolumeUUIDs := make(map[string]string)
+		for _, volume := range volumes {
+			logger.Debug("Creating attached volume snapshot", logger.Ctx{"volumeProject": volume.Project, "volumePool": volume.Pool, "volumeName": volume.Name})
```

Comment:
Lets stick with logging conventions of the fields: `pool`, `volume`, `project`.

Also please use d.logger so we get the instance context.